src/plugins/intel_gpu/src/graph/impls/ocl/kv_cache.cpp

```diff
@@ -0,0 +1,141 @@
+        OPENVINO_THROW("kv_cache axis exceeds number of dimensions");
+
+    // Difference in dimension ordering between IE and GPU plugin,
+    // reverse spatial dimensions after batch and feature.
+    if (cldnn_axis >= 2) {
+        auto spatial_axis = cldnn_axis - 2;
+        // Default and minimum number of dimensions is 4
+        auto spatial_size = std::max<size_t>(rank, 4) - 2;
+        cldnn_axis = spatial_size - spatial_axis - 1 + 2;
+    }
+
+    switch (cldnn_axis) {
+        case 0: return kernel_selector::concat_axis::BATCH;
+        case 1: return kernel_selector::concat_axis::FEATURE;
+        case 2: return kernel_selector::concat_axis::X;
+        case 3: return kernel_selector::concat_axis::Y;
+        case 4: return kernel_selector::concat_axis::Z;
+        case 5: return kernel_selector::concat_axis::W;
+        default: OPENVINO_THROW("Unsupported kv_cache axis: ", axis);
+    }
+
+    return kernel_selector::concat_axis::FEATURE;  // shouldn't get here
+}
+
+}  // namespace
+
+struct kv_cache_impl : typed_primitive_impl_ocl<kv_cache> {
+    using parent = typed_primitive_impl_ocl<kv_cache>;
+    using parent::parent;
+    using kernel_selector_t = kernel_selector::concatenation_kernel_selector;
+    using kernel_params_t = std::pair<kernel_selector::concatenation_params, kernel_selector::concatenation_optional_params>;
+
+    DECLARE_OBJECT_TYPE_SERIALIZATION(cldnn::ocl::kv_cache_impl)
+
+    std::unique_ptr<primitive_impl> clone() const override {
+        return make_unique<kv_cache_impl>(*this);
+    }
+
+    void load(BinaryInputBuffer& ib) override {
+        parent::load(ib);
+        if (is_dynamic()) {
+            auto& kernel_selector = kernel_selector_t::Instance();
+            auto kernel_impl = kernel_selector.GetImplementation(_kernel_data.kernelName);
+            kernel_impl->GetUpdateDispatchDataFunc(_kernel_data);
+        }
+    }
+
+    kernel_arguments_data get_arguments(const kv_cache_inst& instance) const override {
+        kernel_arguments_data args = parent::get_arguments(instance);
+
+        args.inputs = { instance.input_memory_ptr(0), instance.input_memory_ptr(1) };
+
+        return args;
+    }
+
+    event::ptr execute_impl(const std::vector<event::ptr>& events, kv_cache_inst& instance) override {
+        const bool can_be_optimized = instance.get_impl_params()->_can_be_optimized;
+        const auto& desc = instance.get_typed_desc<kv_cache>();
+        auto& variable = instance.get_network().get_variable(desc->variable_info.variable_id);
+        auto res_event = parent::execute_impl(events, instance);
+        variable.set();
+
+        if (can_be_optimized) {
+            // When primitive is optimized, concat kernel writes directly to variable memory
+            return res_event;
+        } else {
+            // Othwerise, we need to copy result from out buffer to state memory
+            auto& stream = instance.get_network().get_stream();
+
+            stream.enqueue_barrier();
```

Comment:
Do we need barrier here? Can `res_event` be passed as dependency for memory copy? Barrier may introduce extra synchronization with other kernels

Reply:
It's possible in theory, but there are 2 issues:
1. There's no API for copy_from/copy_to which handles events for now
2. If we add this API, we'll need to pass additional info (sync_method from ocl_stream impl) to use expected synchronization method

Also, with barrier-based synchronization (which is our default) - that's pretty much expected way to synchronize as event is dummy in that case.